### PR TITLE
[Relay] A Normal Form

### PR DESCRIPTION
**Pull request by MarisaKirisame**
This PR add A Normal Form to relay, which is a common IR form used for subsequent optimization pass.
I had look at the test, and they seems correct. I would like to ensure that with alpha eq, but it is unfinished (some of the case are at dead code elimination), so I would want dce to be merged and I can update the test.
Edit: dce is merged, and I used alpha_equal for test correctness. all is good.

### CODE REVIEW

**Comment by MarisaKirisame** [comment]
@junrushao1994 @srkreddy1238 @siju-samuel can you guys review?

File: include/tvm/relay/pass.h [hunk 1]
```diff
@@ -135,6 +135,12 @@ tvm::Array<TypeParam> FreeTypeVariables(const Type& t);
  */
 Expr DeadCodeElimination(const Expr& e);
 
+/*! \brief Convert the expression into A Normal Form (ANF).
+ *  A Normal Form restrict the program,
+ *  such that all intermediate results are bound to variables via let.
+ */
+Expr ANF(const Expr& e);
```
Comment (tqchen) [inline include/tvm/relay/pass.h]: ANF-> ANormalForm

File: python/tvm/relay/ir_pass.py [hunk 2]
```diff
@@ -69,3 +69,20 @@ def alpha_equal(lhs, rhs):
       True iff lhs is alpha equal to rhs.
     """
     return bool(_make._alpha_equal(lhs, rhs))
+
+def anf(e):
```
Comment (tqchen) [inline python/tvm/relay/ir_pass.py]: a_normal_form

File: src/relay/pass/anf.cc [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+#include "./type_visitor.h"
+#include "tvm/relay/pass.h"
+#include "./let_list.h"
+namespace tvm {
+namespace relay {
+
+class ANFMutator : private ExprMutator {
+  LetList ll;
+  template<typename K, typename V>
+  using Map = std::shared_ptr<std::unordered_map<K, V, NodeHash, NodeEqual> >;
```
Comment (tqchen) [inline src/relay/pass/anf.cc]: shared_ptr not necessary here

File: src/relay/pass/anf.cc [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+ */
+#include <tvm/relay/expr_functor.h>
+#include "./type_visitor.h"
+#include "tvm/relay/pass.h"
+#include "./let_list.h"
+namespace tvm {
+namespace relay {
+
+class ANFMutator : private ExprMutator {
+  LetList ll;
```
Comment (tqchen) [inline src/relay/pass/anf.cc]: explicit mark private member, style, 

```
ll-> let_list_
```

File: src/relay/pass/anf.cc [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+  template<typename K, typename V>
+  using Map = std::shared_ptr<std::unordered_map<K, V, NodeHash, NodeEqual> >;
+  Map<Var, Expr> var_to_expr_;
```
Comment (tqchen) [inline src/relay/pass/anf.cc]: document what each map corresponds to

File: src/relay/pass/anf.cc [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  ANFMutator(const Map<Var, Expr>& var_to_expr,
+             const Map<Var, Var>& var_to_var,
+             const Map<Expr, Var>& expr_to_var) :
+    var_to_expr_(var_to_expr), var_to_var_(var_to_var), expr_to_var_(expr_to_var) { }
+
+  /* while it seems like a good idea to override VisitExpr to push into letlist once for all,
+   * at quite a few case we dont want that (let, param, localvar, globalvar)
+   */
+
+  Expr Compound(const Expr& original, const Expr& e) {
```
Comment (tqchen) [inline src/relay/pass/anf.cc]: document this function

File: src/relay/pass/anf.cc [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+                                 ANF(var_to_expr_, var_to_var_, expr_to_var_, op->true_branch),
+                                 ANF(var_to_expr_, var_to_var_, expr_to_var_, op->false_branch)));
+  }
+
+  Expr VisitExpr_(const OpNode* op) final {
+    return GetRef<Op>(op);
+  }
+
+ public:
+  static Expr ANF(const Map<Var, Expr>& var_to_expr,
```
Comment (tqchen) [inline src/relay/pass/anf.cc]: The additional level of indirection is not necessary, directly inline into caller.

File: src/relay/pass/anf.cc (hunk 3 above)
Comment (MarisaKirisame) [reply src/relay/pass/anf.cc]: it sort of is. when I recurse, I need to pass in all the map, as the recursed is under the same scope.

Comment (MarisaKirisame) [reply src/relay/pass/anf.cc]: (Granted, passing raw map work just as fine, but map is not small, so I will like to share them

Comment (tqchen) [reply src/relay/pass/anf.cc]: likely you can get away by passing pointer

**Comment by tqchen** [comment]
close for now due to inactive status. Feel free to reopen later, A-normal form transformation is important, but need to be aware of the data-flow graph patterns and folds common expression into the same var. Another thing that need to be handled is to pick the right scope of the variable

**Pull request closed without merging**